zap/src/irgen/gen.rs

```diff
@@ -1,517 +1,603 @@
+		NumTy::I32 => buffer_readi32(),
+	}
+}
 
-					if self.ser_checks {
-						self.check_range(&"len".into(), len);
-					}
+fn buffer_readstring(count: Expr) -> Expr {
+	// buffer.readstring(incoming_buff, read(<count>), <count>)
+	Var::from("buffer").nindex("readstring").call(vec![
+		"incoming_buff".into(),
+		Var::from("read").call(vec![count.clone()]),
+		count,
+	])
+}
 
-					self.write_num("len".into(), NumTy::U16);
-					self.write_str(from_expr, "len".into());
+fn range_check(stmts: &mut Vec<Stmt>, val: Expr, range: Range<f64>) {
+	if let Some(min) = range.min() {
+		assert(stmts, val.clone().gte(min.into()), None);
+	}
 
-					self.block_end();
-				}
-			}
+	if let Some(max) = range.max() {
+		assert(
+			stmts,
+			if range.max_inclusive() {
+				val.lte(max.into())
+			} else {
+				val.lt(max.into())
+			},
+			None,
+		);
+	}
+}
 
-			Ty::Arr { len, ty } => {
-				if len.is_exact() {
-					self.num_for("i".into(), 1.into(), len.min().unwrap().into());
+pub fn gen_ser(ty: &Ty, from: Var, gen_checks: bool) -> Vec<Stmt> {
+	let mut stmts = Vec::new();
+	let from_expr = Expr::Var(Box::new(from.clone()));
 
-					self.ser(ty, &from.clone().expr_index("i".into()));
+	if gen_checks {
+		match ty {
+			Ty::F32(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
+			Ty::F64(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
 
-					self.block_end();
-				} else {
-					self.block_start();
+			Ty::U8(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
+			Ty::U16(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
+			Ty::U32(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
 
-					self.local("len");
-					self.assign("len".into(), from_expr.clone().len());
+			Ty::I8(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
+			Ty::I16(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
+			Ty::I32(range) => range_check(&mut stmts, from_expr.clone(), range.cast()),
 
-					if self.ser_checks {
-						self.check_range(&"len".into(), len);
-					}
+			_ => {}
+		};
+	}
 
-					self.write_num("len".into(), NumTy::U16);
+	match ty {
+		Ty::Bool => buffer_writeu8(&mut stmts, from_expr.and(1.0.into()).or(0.0.into())),
 
-					self.num_for("i".into(), 1.into(), "len".into());
+		Ty::F32(..) => buffer_writef32(&mut stmts, from_expr),
+		Ty::F64(..) => buffer_writef64(&mut stmts, from_expr),
 
-					self.ser(ty, &from.clone().expr_index("i".into()));
+		Ty::U8(..) => buffer_writeu8(&mut stmts, from_expr),
+		Ty::U16(..) => buffer_writeu16(&mut stmts, from_expr),
+		Ty::U32(..) => buffer_writeu32(&mut stmts, from_expr),
 
-					self.block_end();
+		Ty::I8(..) => buffer_writei8(&mut stmts, from_expr),
+		Ty::I16(..) => buffer_writei16(&mut stmts, from_expr),
+		Ty::I32(..) => buffer_writei32(&mut stmts, from_expr),
 
-					self.block_end();
+		Ty::Str { len } => {
+			if let Some(len) = len.exact_f64() {
+				if gen_checks {
```

Comment:
This should be moved up scope to line 260 like the rest of the `gen_checks`.

Reply:
I'd rather keep the string checks grouped together as it might cause readers of the code to miss the check for exact strings if it were moved above.